opendevin/memory/history.py

```diff
@@ -1,54 +1,260 @@
+        self.delegates = {}
+
+    def set_event_stream(self, event_stream: EventStream):
+        self._event_stream = event_stream
+
+    def get_events_as_list(self) -> list[Event]:
         """
-        Initialize the empty list of events
+        Return the history as a list of Event objects.
         """
-        self.events = []
+        return list(self.get_events())
 
-    def add_event(self, event_dict: dict):
+    def get_events(self, reverse: bool = False) -> Iterable[Event]:
         """
-        Adds an event to memory if it is a valid event.
+        Return the events as a stream of Event objects.
+        """
+        # TODO handle AgentRejectAction, if it's not part of a chunk ending with an AgentDelegateObservation
+        # or even if it is, because currently we don't add it to the summary
+
+        # iterate from start_id to end_id, or reverse
+        start_id = self.start_id if self.start_id != -1 else 0
+        end_id = (
+            self.end_id
+            if self.end_id != -1
+            else self._event_stream.get_latest_event_id()
+        )
+
+        for event in self._event_stream.get_events(
+            start_id=start_id,
+            end_id=end_id,
+            reverse=reverse,
+            filter_out_type=self.filter_out,
+        ):
+            # TODO add summaries
+            # and filter out events that were included in a summary
 
-        Parameters:
-        - event_dict (dict): The event that we want to add to memory
+            # filter out the events from a delegate of the current agent
+            if not any(
+                # except for the delegate action and observation themselves, currently
+                # AgentDelegateAction has id = delegate_start
+                # AgentDelegateObservation has id = delegate_end
+                delegate_start < event.id < delegate_end
+                for delegate_start, delegate_end in self.delegates.keys()
+            ):
+                yield event
 
-        Raises:
-        - AgentEventTypeError: If event_dict is not a dict
+    def get_last_action(self, end_id: int = -1) -> Action | None:
         """
-        if not isinstance(event_dict, dict):
-            raise AgentEventTypeError()
-        self.events.append(event_dict)
+        Return the last action from the event stream, filtered to exclude unwanted events.
+        """
+        # from end_id in reverse, find the first action
+        end_id = self._event_stream.get_latest_event_id() if end_id == -1 else end_id
+
+        last_action = next(
+            (
+                event
+                for event in self._event_stream.get_events(
+                    end_id=end_id, reverse=True, filter_out_type=self.filter_out
+                )
+                if isinstance(event, Action)
+            ),
+            None,
+        )
 
-    def get_events(self):
+        return last_action
+
+    def get_last_observation(self, end_id: int = -1) -> Observation | None:
```

Comment:
Rename to plural like `get_last_observations` would be clearer, because it _can_ be used for more than one?

Reply:
Same as above, why do you say it returns more than one? Or do you mean, sometimes it doesn't return the same one as other times?

Reply:
No, my misreading after 50 files heh